File: mm-php/src/io/metamessage/core/MM.php
<?php

namespace io\metamessage\core;

use Attribute;
use io\metamessage\ir\ValueType;

#[Attribute(Attribute::TARGET_CLASS | Attribute::TARGET_PROPERTY | Attribute::IS_REPEATABLE)]
class MM
{
    public string $name = '';
    public bool $isNull = false;
    public bool $example = false;
    public bool $deprecated = false;
    public string $desc = '';
    public ValueType $type;
    public bool $nullable = false;
    public bool $allowEmpty = false;
    public bool $unique = false;
    public string $defaultValue = '';
    public string $min = '';
    public string $max = '';
    public int $size = 0;
    public string $enumValues = '';
    public string $pattern = '';
    public int $location = 0;
    public int $version = 0;
    public string $mime = '';

    public string $childDesc = '';
    public ValueType $childType;
    public bool $childNullable = false;
    public bool $childAllowEmpty = false;
    public bool $childUnique = false;
    public string $childDefaultVal = '';
    public string $childMin = '';
    public string $childMax = '';
    public int $childSize = 0;
    public string $childEnums = '';
    public string $childPattern = '';
    public int $childLocation = 0;
    public int $childVersion = 0;
    public string $childMime = '';
    public int $more = 0;

    public function __construct(
        string $name = '',
        bool $isNull = false,
        bool $example = false,
        bool $deprecated = false,
        string $desc = '',
        ValueType $type = ValueType::UNKNOWN,
        bool $nullable = false,
        bool $allowEmpty = false,
        bool $unique = false,
        string $defaultValue = '',
        string $min = '',
        string $max = '',
        int $size = 0,
        string $enumValues = '',
        string $pattern = '',
        int $location = 0,
        int $version = 0,
        string $mime = '',
        string $childDesc = '',
        ValueType $childType = ValueType::UNKNOWN,
        bool $childNullable = false,
        bool $childAllowEmpty = false,
        bool $childUnique = false,
        string $childDefaultVal = '',
        string $childMin = '',
        string $childMax = '',
        int $childSize = 0,
        string $childEnums = '',
        string $childPattern = '',
        int $childLocation = 0,
        int $childVersion = 0,
        string $childMime = '',
        int $more = 0,
    ) {
        $this->name = $name;
        $this->isNull = $isNull;
        $this->example = $example;
        $this->desc = $desc;
        $this->type = $type;
        $this->deprecated = $deprecated;
        $this->nullable = $nullable;
        $this->allowEmpty = $allowEmpty;
        $this->unique = $unique;
        $this->defaultValue = $defaultValue;
        $this->min = $min;
        $this->max = $max;
        $this->size = $size;
        $this->enumValues = $enumValues;
        $this->pattern = $pattern;
        $this->location = $location;
        $this->version = $version;
        $this->mime = $mime;
        $this->childDesc = $childDesc;
        $this->childType = $childType;
        $this->childNullable = $childNullable;
        $this->childAllowEmpty = $childAllowEmpty;
        $this->childUnique = $childUnique;
        $this->childDefaultVal = $childDefaultVal;
        $this->childMin = $childMin;
        $this->childMax = $childMax;
        $this->childSize = $childSize;
        $this->childEnums = $childEnums;
        $this->childPattern = $childPattern;
        $this->childLocation = $childLocation;
        $this->childVersion = $childVersion;
        $this->childMime = $childMime;
        $this->more = $more;
    }
}

File: mm-cs/src/MetaMessage/Core/MM.cs
using MetaMessage.Ir;
using ValueType = MetaMessage.Ir.ValueType;
namespace MetaMessage.Core;

[AttributeUsage(AttributeTargets.Class | AttributeTargets.Property, AllowMultiple = false)]
public class MM : Attribute
{
    public string Name { get; set; } = string.Empty;
    public string Desc { get; set; } = string.Empty;
    public string Type { get; set; } = string.Empty;
    public string ChildType { get; set; } = string.Empty;
    public bool Nullable { get; set; } = false;
    public bool Deprecated { get; set; } = false;
    public bool AllowEmpty { get; set; } = false;
    public bool Unique { get; set; } = false;
    public string DefaultVal { get; set; } = string.Empty;
    public string Enums { get; set; } = string.Empty;
    public int Location { get; set; } = 0;
    public int Version { get; set; } = 0;
    public string Mime { get; set; } = string.Empty;
    public string ChildDesc { get; set; } = string.Empty;
    public bool ChildNullable { get; set; } = false;
    public string ChildEnums { get; set; } = string.Empty;

    public MM()
    {
    }

    public MM(
        string name = "",
        string desc = "",
        string type = "",
        string childType = "",
        bool nullable = false,
        bool deprecated = false,
        bool allowEmpty = false,
        bool unique = false,
        string defaultVal = "",
        string enums = "",
        int location = 0,
        int version = 0,
        string mime = "",
        string childDesc = "",
        bool childNullable = false,
        string childEnums = "")
    {
        Name = name;
        Desc = desc;
        Type = type;
        ChildType = childType;
        Nullable = nullable;
        Deprecated = deprecated;
        AllowEmpty = allowEmpty;
        Unique = unique;
        DefaultVal = defaultVal;
        Enums = enums;
        Location = location;
        Version = version;
        Mime = mime;
        ChildDesc = childDesc;
        ChildNullable = childNullable;
        ChildEnums = childEnums;
    }

    public Tag ToTag()
    {
        var tag = Tag.Empty();

        if (!string.IsNullOrEmpty(Name))
            tag.Name = Name;
        if (!string.IsNullOrEmpty(Desc))
            tag.Desc = Desc;
        if (!string.IsNullOrEmpty(Type))
        {
            if (Enum.TryParse<ValueType>(Type, true, out var type))
                tag.Type = type;
        }
        if (!string.IsNullOrEmpty(ChildType))
        {
            if (Enum.TryParse<ValueType>(ChildType, true, out var childType))
                tag.ChildType = childType;
        }
        tag.Nullable = Nullable;
        tag.Deprecated = Deprecated;
        tag.AllowEmpty = AllowEmpty;
        tag.Unique = Unique;
        if (!string.IsNullOrEmpty(DefaultVal))
            tag.DefaultVal = DefaultVal;
        if (!string.IsNullOrEmpty(Enums))
            tag.Enums = Enums;
        tag.Location = Location;
        tag.Version = Version;
        if (!string.IsNullOrEmpty(Mime))
            tag.Mime = Mime;
        if (!string.IsNullOrEmpty(ChildDesc))
            tag.ChildDesc = ChildDesc;
        tag.ChildNullable = ChildNullable;
        if (!string.IsNullOrEmpty(ChildEnums))
            tag.ChildEnums = ChildEnums;

        return tag;
    }

    public static MM FromTag(Tag tag)
    {
        return new MM(
            name: tag.Name,
            desc: tag.Desc,
            type: tag.Type != ValueType.Unknown ? tag.Type.ToString() : "",
            childType: tag.ChildType != ValueType.Unknown ? tag.ChildType.ToString() : "",
            nullable: tag.Nullable,
            deprecated: tag.Deprecated,
            allowEmpty: tag.AllowEmpty,
            unique: tag.Unique,
            defaultVal: tag.DefaultVal,
            enums: tag.Enums,
            location: tag.Location,
            version: tag.Version,
            mime: tag.Mime,
            childDesc: tag.ChildDesc,
            childNullable: tag.ChildNullable,
            childEnums: tag.ChildEnums);
    }
}

public static class TagExtensions
{
    public static Tag FromAttribute(MM attribute)
    {
        return attribute.ToTag();
    }
}